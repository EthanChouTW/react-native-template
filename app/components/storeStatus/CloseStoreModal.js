import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Modal, Platform, StyleSheet, Text, TextInput, TouchableOpacity, View } from 'react-native';
import RadioButton from '../buttons/RadioButton';
import RadioGroup from '../buttons/RadioGroup';
import { DurationTypes } from '../../utils/appConstants';
import Theme from '../../utils/styleCollection';
import Locales from '../../locales';

const DEFAULT_MINUTES_DURATION = 20;
const DEFAULT_HOURS_DURATION = 1;
const DAYS_DURATION_TODAY = 1;
const DEFAULT_DAYS_INPUT = '2';

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: Theme.modalGreyOut,
    justifyContent: 'center'
  },
  dialogContainer: {
    padding: 24,
    marginHorizontal: 40,
    backgroundColor: 'white'
  },
  title: {
    fontSize: 20,
    color: Theme.blackTwo54,
    marginBottom: 16
  },
  radioGroup: {
    marginLeft: -8
  },
  radioButtonLabel: {
    fontSize: 16,
    color: 'black',
    paddingLeft: 12
  },
  daysRadioButton: {
    marginTop: -10,
    alignItems: 'center'
  },
  daysTextInputContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    marginLeft: 8.5,
    borderBottomWidth: Platform.OS === 'ios' ? 1 : 0,
    borderColor: Theme.shamrockGreen,
    marginRight: Platform.OS === 'ios' ? 10 : 0
  },
  daysTextInput: {
    height: 40,
    width: '100%',
    fontSize: 16,
    color: 'black'
  },
  daysLabel: {
    position: 'absolute',
    right: Platform.OS === 'ios' ? 0 : 10,
    bottom: 10,
    fontSize: 16,
    color: Theme.darkerGray
  },
  buttonContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'flex-end'
  },
  buttonText: {
    fontSize: 16,
    marginTop: 8,
    color: Theme.shamrockGreen,
    fontWeight: 'bold',
    marginLeft: 20
  }
});

const defaultState = {
  selectedIndex: 0,
  saveBtnDisabled: false,
  daysInput: DEFAULT_DAYS_INPUT
};

export default class CloseStoreModal extends Component {
  static propTypes = {
    visible: PropTypes.bool,
    onCancel: PropTypes.func,
    onSave: PropTypes.func
  };

  constructor(props) {
    super(props);
    this.state = defaultState;
  }

  onSelect = index => {
    this.setState({ selectedIndex: index });

    if (index !== 3) {
      this.daysTextInputRef.blur();
    }
  };

  onDaysChang = daysInput => {
    const days = parseInt(daysInput, 10);

    if (!isNaN(days) && days >= 1 && days <= 7) {
      this.setState({ saveBtnDisabled: false, daysInput });
    } else {
      this.setState({ saveBtnDisabled: true, daysInput: '' });
    }
  };

  onSave = () => {
    let durationType;
    let duration;

    switch (this.state.selectedIndex) {
      default:
      case 0:
        durationType = DurationTypes.MINUTES;
        duration = DEFAULT_MINUTES_DURATION;
        break;
      case 1:
        durationType = DurationTypes.HOURS;
        duration = DEFAULT_HOURS_DURATION;
        break;
      case 2:
        durationType = DurationTypes.DAYS;
        duration = DAYS_DURATION_TODAY;
        break;
      case 3:
        durationType = DurationTypes.DAYS;
        duration = parseInt(this.state.daysInput, 10);
        break;
    }

    this.props.onSave(durationType, duration);
  };

  render() {
    return (
      <Modal
        animationType="slide"
        transparent
        visible={this.props.visible}
        onShow={() => this.setState(defaultState)}
        onRequestClose={this.props.onCancel}
      >
        <View style={styles.container}>
          <View style={styles.dialogContainer}>
            <Text style={styles.title}>{Locales.t('close_store_dialog_title')}</Text>

            <RadioGroup
              style={styles.radioGroup}
              color={Theme.darkerGray}
              activeColor={Theme.shamrockGreen}
              selectedIndex={this.state.selectedIndex}
              onSelect={(index, value) => this.onSelect(index, value)}
            >
              <RadioButton>
                <Text style={styles.radioButtonLabel}>
                  {Locales.t('close_store_duration_minutes')}
                </Text>
              </RadioButton>

              <RadioButton>
                <Text style={styles.radioButtonLabel}>
                  {Locales.t('close_store_duration_hours')}
                </Text>
              </RadioButton>

              <RadioButton>
                <Text style={styles.radioButtonLabel}>{Locales.t('today')}</Text>
              </RadioButton>

              <RadioButton style={styles.daysRadioButton}>
                <View style={styles.daysTextInputContainer}>
                  <TextInput
                    ref={ref => (this.daysTextInputRef = ref)}
                    style={styles.daysTextInput}
                    underlineColorAndroid={Theme.shamrockGreen}
                    onChangeText={this.onDaysChang}
                    keyboardType={'numeric'}
                    maxLength={1}
                    value={this.state.daysInput}
                    onFocus={() => this.setState({ selectedIndex: 3 })}
                  />
                  <Text style={styles.daysLabel}>{Locales.t('days')}</Text>
                </View>
              </RadioButton>
            </RadioGroup>

            <View style={styles.buttonContainer}>
              <TouchableOpacity onPress={this.props.onCancel}>
                <Text style={styles.buttonText}>{Locales.t('lbl_cancel')}</Text>
              </TouchableOpacity>

              <TouchableOpacity onPress={this.onSave} disabled={this.state.saveBtnDisabled}>
                <Text
                  style={[
                    styles.buttonText,
                    this.state.saveBtnDisabled && { color: Theme.disabledGreen }
                  ]}
                >
                  {Locales.t('lbl_save')}
                </Text>
              </TouchableOpacity>
            </View>
          </View>
        </View>
      </Modal>
    );
  }
}
