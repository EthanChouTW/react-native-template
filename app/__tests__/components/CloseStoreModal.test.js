import 'react-native';
import React from 'react';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import CloseStoreModal from '../../components/storeStatus/CloseStoreModal';

describe('CloseStoreModal', () => {
  const props = {
    visible: true,
    onCancel: jest.fn(),
    onSave: jest.fn()
  };

  test('renders correctly', () => {
    const snapshot = renderer.create(<CloseStoreModal {...props} />).toJSON();
    expect(snapshot).toMatchSnapshot();
  });
});
