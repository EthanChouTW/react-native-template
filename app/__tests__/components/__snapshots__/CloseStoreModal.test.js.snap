// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CloseStoreModal renders correctly 1`] = `
<Modal
  animationType="slide"
  hardwareAccelerated={false}
  onRequestClose={[Function]}
  onShow={[Function]}
  transparent={true}
  visible={true}
>
  <
    style={
      Object {
        "backgroundColor": "#00000080",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <
      style={
        Object {
          "backgroundColor": "white",
          "marginHorizontal": 40,
          "padding": 24,
        }
      }
    >
      <
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#000000D4",
            "fontSize": 20,
            "marginBottom": 16,
          }
        }
      >
        close_store_dialog_title
      </>
      <
        style={
          Object {
            "marginLeft": -8,
          }
        }
      >
        <
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <
            accessibilityComponentType={undefined}
            accessibilityLabel={undefined}
            accessibilityTraits={undefined}
            accessible={true}
            hitSlop={undefined}
            nativeID={undefined}
            onLayout={undefined}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "flexDirection": "row",
                  "flexGrow": 1,
                  "padding": 10,
                },
                undefined,
                Object {
                  "backgroundColor": null,
                },
              ]
            }
            testID={undefined}
          >
            <
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "justifyContent": "center",
                  },
                  Object {
                    "borderColor": "#07c748",
                    "borderRadius": 10,
                    "borderWidth": 1,
                    "height": 20,
                    "width": 20,
                  },
                ]
              }
            >
              <
                style={
                  Object {
                    "backgroundColor": "#07c748",
                    "borderRadius": 5,
                    "height": 10,
                    "width": 10,
                  }
                }
              />
            </>
            <
              style={
                Object {
                  "alignItems": "center",
                  "justifyContent": "center",
                  "marginLeft": 5,
                }
              }
            >
              <
                accessible={true}
                allowFontScaling={true}
                disabled={false}
                ellipsizeMode="tail"
                style={
                  Object {
                    "color": "black",
                    "fontSize": 16,
                    "paddingLeft": 12,
                  }
                }
              >
                close_store_duration_minutes
              </>
            </>
          </>
        </>
        <
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <
            accessibilityComponentType={undefined}
            accessibilityLabel={undefined}
            accessibilityTraits={undefined}
            accessible={true}
            hitSlop={undefined}
            nativeID={undefined}
            onLayout={undefined}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "flexDirection": "row",
                  "flexGrow": 1,
                  "padding": 10,
                },
                undefined,
                null,
              ]
            }
            testID={undefined}
          >
            <
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "justifyContent": "center",
                  },
                  Object {
                    "borderColor": "#757575",
                    "borderRadius": 10,
                    "borderWidth": 1,
                    "height": 20,
                    "width": 20,
                  },
                ]
              }
            >
              <
                style={
                  Object {
                    "borderColor": "#757575",
                    "borderRadius": 10,
                    "borderWidth": 1,
                    "height": 20,
                    "width": 20,
                  }
                }
              />
            </>
            <
              style={
                Object {
                  "alignItems": "center",
                  "justifyContent": "center",
                  "marginLeft": 5,
                }
              }
            >
              <
                accessible={true}
                allowFontScaling={true}
                disabled={false}
                ellipsizeMode="tail"
                style={
                  Object {
                    "color": "black",
                    "fontSize": 16,
                    "paddingLeft": 12,
                  }
                }
              >
                close_store_duration_hours
              </>
            </>
          </>
        </>
        <
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <
            accessibilityComponentType={undefined}
            accessibilityLabel={undefined}
            accessibilityTraits={undefined}
            accessible={true}
            hitSlop={undefined}
            nativeID={undefined}
            onLayout={undefined}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "flexDirection": "row",
                  "flexGrow": 1,
                  "padding": 10,
                },
                undefined,
                null,
              ]
            }
            testID={undefined}
          >
            <
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "justifyContent": "center",
                  },
                  Object {
                    "borderColor": "#757575",
                    "borderRadius": 10,
                    "borderWidth": 1,
                    "height": 20,
                    "width": 20,
                  },
                ]
              }
            >
              <
                style={
                  Object {
                    "borderColor": "#757575",
                    "borderRadius": 10,
                    "borderWidth": 1,
                    "height": 20,
                    "width": 20,
                  }
                }
              />
            </>
            <
              style={
                Object {
                  "alignItems": "center",
                  "justifyContent": "center",
                  "marginLeft": 5,
                }
              }
            >
              <
                accessible={true}
                allowFontScaling={true}
                disabled={false}
                ellipsizeMode="tail"
                style={
                  Object {
                    "color": "black",
                    "fontSize": 16,
                    "paddingLeft": 12,
                  }
                }
              >
                today
              </>
            </>
          </>
        </>
        <
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <
            accessibilityComponentType={undefined}
            accessibilityLabel={undefined}
            accessibilityTraits={undefined}
            accessible={true}
            hitSlop={undefined}
            nativeID={undefined}
            onLayout={undefined}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "flexDirection": "row",
                  "flexGrow": 1,
                  "padding": 10,
                },
                Object {
                  "alignItems": "center",
                  "marginTop": -10,
                },
                null,
              ]
            }
            testID={undefined}
          >
            <
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "justifyContent": "center",
                  },
                  Object {
                    "borderColor": "#757575",
                    "borderRadius": 10,
                    "borderWidth": 1,
                    "height": 20,
                    "width": 20,
                  },
                ]
              }
            >
              <
                style={
                  Object {
                    "borderColor": "#757575",
                    "borderRadius": 10,
                    "borderWidth": 1,
                    "height": 20,
                    "width": 20,
                  }
                }
              />
            </>
            <
              style={
                Object {
                  "alignItems": "center",
                  "justifyContent": "center",
                  "marginLeft": 5,
                }
              }
            >
              <
                style={
                  Object {
                    "alignItems": "center",
                    "borderBottomWidth": 1,
                    "borderColor": "#07c748",
                    "flexDirection": "row",
                    "marginLeft": 8.5,
                    "marginRight": 10,
                  }
                }
              >
                <
                  keyboardType="numeric"
                  maxLength={1}
                  onChangeText={[Function]}
                  onFocus={[Function]}
                  style={
                    Object {
                      "color": "black",
                      "fontSize": 16,
                      "height": 40,
                      "width": "100%",
                    }
                  }
                  underlineColorAndroid="#07c748"
                  value="2"
                />
                <
                  accessible={true}
                  allowFontScaling={true}
                  disabled={false}
                  ellipsizeMode="tail"
                  style={
                    Object {
                      "bottom": 10,
                      "color": "#757575",
                      "fontSize": 16,
                      "position": "absolute",
                      "right": 0,
                    }
                  }
                >
                  days
                </>
              </>
            </>
          </>
        </>
      </>
      <
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "flex-end",
          }
        }
      >
        <
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          collapsable={undefined}
          hitSlop={undefined}
          isTVSelectable={true}
          nativeID={undefined}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <
            accessible={true}
            allowFontScaling={true}
            disabled={false}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#07c748",
                "fontSize": 16,
                "fontWeight": "bold",
                "marginLeft": 20,
                "marginTop": 8,
              }
            }
          >
            lbl_cancel
          </>
        </>
        <
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          collapsable={undefined}
          hitSlop={undefined}
          isTVSelectable={true}
          nativeID={undefined}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <
            accessible={true}
            allowFontScaling={true}
            disabled={false}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#07c748",
                  "fontSize": 16,
                  "fontWeight": "bold",
                  "marginLeft": 20,
                  "marginTop": 8,
                },
                false,
              ]
            }
          >
            lbl_save
          </>
        </>
      </>
    </>
  </>
</Modal>
`;
